File: leetcode-04/solutions/cpp/0046.cpp
class Solution {
 public:
  vector<vector<int>> permute(vector<int>& nums) {
    vector<vector<int>> ans;
    vector<int> path;
    vector<bool> used(nums.size(), false);

    dfs(nums, nums.size(), used, path, ans);

    return ans;
  }

 private:
  void dfs(vector<int>& nums, int target, vector<bool>& used, vector<int>& path,
           vector<vector<int>>& ans) {
    if (target == 0) {
      ans.push_back(path);
      return;
    }

    for (int i = 0; i < nums.size(); i++) {
      if (used[i]) continue;
      used[i] = true;
      path.push_back(nums[i]);
      dfs(nums, target - 1, used, path, ans);
      path.pop_back();
      used[i] = false;
    }
  }
};

File: leetcode-04/solutions/python3/0046.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []

        self.dfs(nums, len(nums), [False] * len(nums), [], ans)

        return ans

    def dfs(self, nums: List[int], target: int, used: List[bool], path: List[int], ans: List[List[int]]) -> None:
        if target == 0:
            ans.append(path)
            return

        for i in range(len(nums)):
            if used[i]:
                continue
            used[i] = True
            self.dfs(nums, target - 1, used, path + [nums[i]], ans)
            used[i] = False
